# This notebook will be used to run the ML algorithms.

First let us import the necessary libraries.

In [21]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Next we have to read the dataset.

In [22]:
df = pd.read_csv("adult.csv")
df

age workclass  fnlwgt     education  education.num      marital.status  \
0       90         ?   77053       HS-grad              9             Widowed   
1       82   Private  132870       HS-grad              9             Widowed   
2       66         ?  186061  Some-college             10             Widowed   
3       54   Private  140359       7th-8th              4            Divorced   
4       41   Private  264663  Some-college             10           Separated   
...    ...       ...     ...           ...            ...                 ...   
32556   22   Private  310152  Some-college             10       Never-married   
32557   27   Private  257302    Assoc-acdm             12  Married-civ-spouse   
32558   40   Private  154374       HS-grad              9  Married-civ-spouse   
32559   58   Private  151910       HS-grad              9             Widowed   
32560   22   Private  201490       HS-grad              9       Never-married   

              occupation   relationship   race     sex  capital.gain  \
0                      ?  Not-in-family  White  Female             0   
1        Exec-managerial  Not-in-family  White  Female             0   
2                      ?      Unmarried  Black  Female             0   
3      Machine-op-inspct      Unmarried  White  Female             0   
4         Prof-specialty      Own-child  White  Female             0   
...                  ...            ...    ...     ...           ...   
32556    Protective-serv  Not-in-family  White    Male             0   
32557       Tech-support           Wife  White  Female             0   
32558  Machine-op-inspct        Husband  White    Male             0   
32559       Adm-clerical      Unmarried  White  Female             0   
32560       Adm-clerical      Own-child  White    Male             0   

       capital.loss  hours.per.week native.country income  
0              4356              40  United-States  <=50K  
1              4356              18  United-States  <=50K  
2              4356              40  United-States  <=50K  
3              3900              40  United-States  <=50K  
4              3900              40  United-States  <=50K  
...             ...             ...            ...    ...  
32556             0              40  United-States  <=50K  
32557             0              38  United-States  <=50K  
32558             0              40  United-States   >50K  
32559             0              40  United-States  <=50K  
32560             0              20  United-States  <=50K  

[32561 rows x 15 columns]

It could be seen that our dataset has 32561 rows and 15 columns (features).

Let us check the dataset for null values.

In [23]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64

Upon careful observation, it could be seen that though the code provides that there are no null values, some of the values in the columns with data type have been filled with just "?".

Rows which have such entries could be omitted from the data frame.

In [24]:
df[df=='?'] = np.nan # if there is a question mark, it is filled with NaN (null)
df = df.dropna(axis=0) # rows with a null value with atleast one column are dropped
df

age workclass  fnlwgt     education  education.num      marital.status  \
1       82   Private  132870       HS-grad              9             Widowed   
3       54   Private  140359       7th-8th              4            Divorced   
4       41   Private  264663  Some-college             10           Separated   
5       34   Private  216864       HS-grad              9            Divorced   
6       38   Private  150601          10th              6           Separated   
...    ...       ...     ...           ...            ...                 ...   
32556   22   Private  310152  Some-college             10       Never-married   
32557   27   Private  257302    Assoc-acdm             12  Married-civ-spouse   
32558   40   Private  154374       HS-grad              9  Married-civ-spouse   
32559   58   Private  151910       HS-grad              9             Widowed   
32560   22   Private  201490       HS-grad              9       Never-married   

              occupation   relationship   race     sex  capital.gain  \
1        Exec-managerial  Not-in-family  White  Female             0   
3      Machine-op-inspct      Unmarried  White  Female             0   
4         Prof-specialty      Own-child  White  Female             0   
5          Other-service      Unmarried  White  Female             0   
6           Adm-clerical      Unmarried  White    Male             0   
...                  ...            ...    ...     ...           ...   
32556    Protective-serv  Not-in-family  White    Male             0   
32557       Tech-support           Wife  White  Female             0   
32558  Machine-op-inspct        Husband  White    Male             0   
32559       Adm-clerical      Unmarried  White  Female             0   
32560       Adm-clerical      Own-child  White    Male             0   

       capital.loss  hours.per.week native.country income  
1              4356              18  United-States  <=50K  
3              3900              40  United-States  <=50K  
4              3900              40  United-States  <=50K  
5              3770              45  United-States  <=50K  
6              3770              40  United-States  <=50K  
...             ...             ...            ...    ...  
32556             0              40  United-States  <=50K  
32557             0              38  United-States  <=50K  
32558             0              40  United-States   >50K  
32559             0              40  United-States  <=50K  
32560             0              20  United-States  <=50K  

[30162 rows x 15 columns]

Now it is visible that the number of rows has decreased from 32561 to 30162.

The classification problem requires us to predict whether the income exceeds $50,000 per year. So if the income value is '<=50K', then the value is replaced with 0, else if the income value is '>50K', then the value is replaced with 1.

In [26]:
# the values are replaced with the respective new values in the income column
df['income'].replace({'<=50K':0, '>50K':1}, inplace=True)
df

C:\Users\arkha\AppData\Local\Temp\ipykernel_13292\2647791319.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['income'].replace({'<=50K':0, '>50K':1}, inplace=True)


age workclass  fnlwgt     education  education.num      marital.status  \
1       82   Private  132870       HS-grad              9             Widowed   
3       54   Private  140359       7th-8th              4            Divorced   
4       41   Private  264663  Some-college             10           Separated   
5       34   Private  216864       HS-grad              9            Divorced   
6       38   Private  150601          10th              6           Separated   
...    ...       ...     ...           ...            ...                 ...   
32556   22   Private  310152  Some-college             10       Never-married   
32557   27   Private  257302    Assoc-acdm             12  Married-civ-spouse   
32558   40   Private  154374       HS-grad              9  Married-civ-spouse   
32559   58   Private  151910       HS-grad              9             Widowed   
32560   22   Private  201490       HS-grad              9       Never-married   

              occupation   relationship   race     sex  capital.gain  \
1        Exec-managerial  Not-in-family  White  Female             0   
3      Machine-op-inspct      Unmarried  White  Female             0   
4         Prof-specialty      Own-child  White  Female             0   
5          Other-service      Unmarried  White  Female             0   
6           Adm-clerical      Unmarried  White    Male             0   
...                  ...            ...    ...     ...           ...   
32556    Protective-serv  Not-in-family  White    Male             0   
32557       Tech-support           Wife  White  Female             0   
32558  Machine-op-inspct        Husband  White    Male             0   
32559       Adm-clerical      Unmarried  White  Female             0   
32560       Adm-clerical      Own-child  White    Male             0   

       capital.loss  hours.per.week native.country  income  
1              4356              18  United-States       0  
3              3900              40  United-States       0  
4              3900              40  United-States       0  
5              3770              45  United-States       0  
6              3770              40  United-States       0  
...             ...             ...            ...     ...  
32556             0              40  United-States       0  
32557             0              38  United-States       0  
32558             0              40  United-States       1  
32559             0              40  United-States       0  
32560             0              20  United-States       0  

[30162 rows x 15 columns]

It could be noted that the column "education.num" could prove to be useless in this particular classification problem. So this column could be dropped.

In [27]:
df = df.drop('education.num', axis=1) # the column is dropped
df

age workclass  fnlwgt     education      marital.status  \
1       82   Private  132870       HS-grad             Widowed   
3       54   Private  140359       7th-8th            Divorced   
4       41   Private  264663  Some-college           Separated   
5       34   Private  216864       HS-grad            Divorced   
6       38   Private  150601          10th           Separated   
...    ...       ...     ...           ...                 ...   
32556   22   Private  310152  Some-college       Never-married   
32557   27   Private  257302    Assoc-acdm  Married-civ-spouse   
32558   40   Private  154374       HS-grad  Married-civ-spouse   
32559   58   Private  151910       HS-grad             Widowed   
32560   22   Private  201490       HS-grad       Never-married   

              occupation   relationship   race     sex  capital.gain  \
1        Exec-managerial  Not-in-family  White  Female             0   
3      Machine-op-inspct      Unmarried  White  Female             0   
4         Prof-specialty      Own-child  White  Female             0   
5          Other-service      Unmarried  White  Female             0   
6           Adm-clerical      Unmarried  White    Male             0   
...                  ...            ...    ...     ...           ...   
32556    Protective-serv  Not-in-family  White    Male             0   
32557       Tech-support           Wife  White  Female             0   
32558  Machine-op-inspct        Husband  White    Male             0   
32559       Adm-clerical      Unmarried  White  Female             0   
32560       Adm-clerical      Own-child  White    Male             0   

       capital.loss  hours.per.week native.country  income  
1              4356              18  United-States       0  
3              3900              40  United-States       0  
4              3900              40  United-States       0  
5              3770              45  United-States       0  
6              3770              40  United-States       0  
...             ...             ...            ...     ...  
32556             0              40  United-States       0  
32557             0              38  United-States       0  
32558             0              40  United-States       1  
32559             0              40  United-States       0  
32560             0              20  United-States       0  

[30162 rows x 14 columns]

Now we have to check which features of the dataset exactly contribute to the prediction.

Point Biserial Correlation is used for identifying the correlation between a binary variable (income) and a continuous variable.

In [28]:
import scipy.stats as stats
columns = ['age','capital.loss','capital.gain','hours.per.week','fnlwgt'] # columns to be analysed are put in an array
for column in columns:
    pbc_value = stats.pointbiserialr(df['income'], df[column])[0] # the correlation value is selected.
    # if the value is greater than zero, it is positive or else it is negative correlation
    if pbc_value > 0:
        status = "positive correlation"
    else:
        status = "negative correlation"
    print(column, ":", pbc_value, " - ", status) # the column name and the value is printed

age : 0.24199813626610106  -  positive correlation
capital.loss : 0.15005330839725733  -  positive correlation
capital.gain : 0.2211962145480077  -  positive correlation
hours.per.week : 0.22948012988848634  -  positive correlation
fnlwgt : -0.00895742335917096  -  negative correlation


It could be seen that the column "fnlgwt" has a negative correlation with respect to the income column. So it could be dropped from the feature selection.

In [29]:
df = df.drop('fnlwgt', axis=1)
df

age workclass     education      marital.status         occupation  \
1       82   Private       HS-grad             Widowed    Exec-managerial   
3       54   Private       7th-8th            Divorced  Machine-op-inspct   
4       41   Private  Some-college           Separated     Prof-specialty   
5       34   Private       HS-grad            Divorced      Other-service   
6       38   Private          10th           Separated       Adm-clerical   
...    ...       ...           ...                 ...                ...   
32556   22   Private  Some-college       Never-married    Protective-serv   
32557   27   Private    Assoc-acdm  Married-civ-spouse       Tech-support   
32558   40   Private       HS-grad  Married-civ-spouse  Machine-op-inspct   
32559   58   Private       HS-grad             Widowed       Adm-clerical   
32560   22   Private       HS-grad       Never-married       Adm-clerical   

        relationship   race     sex  capital.gain  capital.loss  \
1      Not-in-family  White  Female             0          4356   
3          Unmarried  White  Female             0          3900   
4          Own-child  White  Female             0          3900   
5          Unmarried  White  Female             0          3770   
6          Unmarried  White    Male             0          3770   
...              ...    ...     ...           ...           ...   
32556  Not-in-family  White    Male             0             0   
32557           Wife  White  Female             0             0   
32558        Husband  White    Male             0             0   
32559      Unmarried  White  Female             0             0   
32560      Own-child  White    Male             0             0   

       hours.per.week native.country  income  
1                  18  United-States       0  
3                  40  United-States       0  
4                  40  United-States       0  
5                  45  United-States       0  
6                  40  United-States       0  
...               ...            ...     ...  
32556              40  United-States       0  
32557              38  United-States       0  
32558              40  United-States       1  
32559              40  United-States       0  
32560              20  United-States       0  

[30162 rows x 13 columns]